src/main/java/com/google/api/generator/engine/ast/UnaryOperationExpr.java

```diff
@@ -0,0 +1,102 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class UnaryOperationExpr implements OperationExpr {
+
+  public abstract Expr expr();
+
+  public abstract OperatorKind operatorKind();
+
+  public abstract TypeNode type();
+
+  public void accept(AstNodeVisitor visitor) {
+    visitor.visit(this);
+  }
+
+  public static UnaryOperationExpr postfixIncrementWithExpr(Expr expr) {
+    return builder()
+        .setExpr(expr)
+        .setOperatorKind(OperatorKind.UNARY_POST_INCREMENT)
+        .setType(expr.type())
+        .build();
+  }
+
+  public static UnaryOperationExpr logicalNotWithExpr(Expr expr) {
+    return builder()
+        .setOperatorKind(OperatorKind.UNARY_LOGICAL_NOT)
+        .setExpr(expr)
+        .setType(TypeNode.BOOLEAN)
+        .build();
+  }
+
+  public static Builder builder() {
+    return new AutoValue_UnaryOperationExpr.Builder();
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+
+    // private setter.
+    abstract Builder setExpr(Expr expr);
+
+    // private setter.
+    abstract Builder setOperatorKind(OperatorKind operator);
+
+    // private setter.
+    abstract Builder setType(TypeNode type);
+
+    abstract UnaryOperationExpr autoBuild();
+
+    private UnaryOperationExpr build() {
+      UnaryOperationExpr unaryOperationExpr = autoBuild();
+      TypeNode exprType = unaryOperationExpr.expr().type();
+      OperatorKind operator = unaryOperationExpr.operatorKind();
+      final String errorMsg =
+          String.format(
+              "Unary operator %s can not be applied to %s. ", operator, exprType.toString());
+
+      // Unary Operators can not be applied on Void type
+      Preconditions.checkState(
+          !exprType.equals(TypeNode.VOID) && !exprType.equals(TypeNode.NULL), errorMsg);
+
+      if (operator.equals(OperatorKind.UNARY_LOGICAL_NOT)) {
+        Preconditions.checkState(isValidLogicalNotType(exprType), errorMsg);
+      }
+
+      if (operator.equals(OperatorKind.UNARY_POST_INCREMENT)) {
+        Preconditions.checkState(isValidIncrementType(exprType), errorMsg);
+      }
+
+      return unaryOperationExpr;
+    }
+  }
+
+  private static boolean isValidLogicalNotType(TypeNode exprType) {
+    // Logical not (!) can only be applied on boolean/Boolean type
+    return exprType.equals(TypeNode.BOOLEAN) || exprType.equals(TypeNode.BOOLEAN_OBJECT);
```

Comment:
Just checking for `boolean` should be sufficient, since the `equals()` implementation takes care of the boxed type.

Reply:
sg